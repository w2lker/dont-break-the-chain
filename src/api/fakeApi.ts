import profile from './fakeData/profile';

const fakeFetch = (data: any) => {
  return new Promise((resolve, reject) => {
    if (data) {
      setTimeout(() => {
        resolve({
          data,
        });
      }, 200);
    } else {
      reject({
        error: 'Some error message',
      });
    }
  });
};

class Api {
  getProfile() {
    return fakeFetch(profile);
  }
  logout() {
    return fakeFetch({ status: 200 });
  }
}

export default new Api();
