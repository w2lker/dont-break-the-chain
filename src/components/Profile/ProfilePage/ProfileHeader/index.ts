import ProfileHeader from './ProfileHeader.decorators';

export default ProfileHeader;
