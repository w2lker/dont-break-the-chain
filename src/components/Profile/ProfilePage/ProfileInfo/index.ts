import ProfileInfo from './ProfileInfo.decorators';

export default ProfileInfo;
