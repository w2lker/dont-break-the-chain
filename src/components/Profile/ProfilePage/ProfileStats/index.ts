import ProfileStats from './ProfileStats.decorators';

export default ProfileStats;
