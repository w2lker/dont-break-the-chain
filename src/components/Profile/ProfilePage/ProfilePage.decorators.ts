import { Dispatch } from 'react';
import { connect } from 'react-redux';
import { getProfile } from '../../../actions/profile';
import { IProfileReducer } from '../../../models/profile';
import ProfilePage, { IProfilePageProps } from './ProfilePage';

const mapStateToProps = (state: any): IProfilePageProps => {
  const profile: IProfileReducer = state.profile;
  // @ts-ignore
  return {
    id: profile.id,
  };
};

const mapDispatchToProps = (dispatch: Dispatch<any>): IProfilePageProps => {
  return {
    getProfile: () => dispatch(getProfile()),
  };
};

const connectedComponent = connect(mapStateToProps, mapDispatchToProps)(ProfilePage);

export default connectedComponent;
