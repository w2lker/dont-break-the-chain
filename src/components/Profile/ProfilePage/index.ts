import ProfilePage from './ProfilePage.decorators';

export default ProfilePage;
