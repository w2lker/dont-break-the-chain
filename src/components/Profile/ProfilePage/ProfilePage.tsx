import React from 'react';
import ProfileHeader from './ProfileHeader';
import ProfileInfo from './ProfileInfo';
import ProfileStats from './ProfileStats';

export interface IProfilePageProps {
  id?: number;
  profileLoading?: boolean;
  profileLoadError?: boolean;
  getProfile: () => void;
}

class ProfilePage extends React.Component<IProfilePageProps, {}> {

  componentDidMount(): void {
    const { id, getProfile } = this.props;
    if (!id) {
      getProfile();
    }
  }

  render(): React.ReactElement<any, string | React.JSXElementConstructor<any>> | string | number | {} | React.ReactNodeArray | React.ReactPortal | boolean | null | undefined {
    const { id } = this.props;
    // TODO: provide loading and load error states to this component
    if (!id) {
      return null;
    }
    return (
      <React.Fragment>
        <ProfileHeader />
        <ProfileInfo />
        <ProfileStats />
      </React.Fragment>
    );
  }
}

export default ProfilePage;
