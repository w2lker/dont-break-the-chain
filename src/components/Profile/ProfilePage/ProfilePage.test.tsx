import React from 'react';

import { mount, shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import fakeApi from '../../../api/fakeApi';
import ProfilePage, { IProfilePageProps } from './ProfilePage';
import ProfilePageDecorated from './ProfilePage.decorators';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

describe('ProfilePage component', () => {
  const sampleProps: IProfilePageProps = {
    id: 123,
    getProfile: jest.fn(),
  };

  const sampleStore = {
    profile: {
      id: 3423,
      name: 'Test User',
      avatar: './img.png',
      email: 'tester@gmail.com',
      points: 245,
      longestChain: 12,
    },
  };

  beforeEach(() => {
    // @ts-ignore
    sampleProps.getProfile.mockReset();
  });

  it('matches snapshot', () => {
    const store = mockStore(sampleStore);
    const component = mount(
      <Provider store={store}>
        <ProfilePage {...sampleProps} />
      </Provider>,
    );
    expect(component.debug()).toMatchSnapshot();
  });

  it('skips render without id', () => {
    const props = {
      ...sampleProps,
      id: null,
    };
    // @ts-ignore
    const component = shallow(<ProfilePage {...props} />);
    expect(component).toEqual({});
    expect(sampleProps.getProfile).toBeCalledTimes(1);
  });

  it('renders with props', () => {
    const store = mockStore(sampleStore);
    const component = mount(
      <Provider store={store}>
        <ProfilePage {...sampleProps} />);
      </Provider>,
    );
    expect(component.find('ProfileHeader').length).toBe(1);
    expect(component.find('ProfileInfo').length).toBe(1);
    expect(component.find('ProfileStats').length).toBe(1);
    expect(sampleProps.getProfile).toBeCalledTimes(0);
  });
});

describe('ProfilePage decorators', () => {

  it('provides id from connected store', () => {
    const storeWithID = {
      profile: {
        id: 14532,
      },
    };
    const store = mockStore(storeWithID);
    const component = mount(
      <Provider store={store}>
        <ProfilePageDecorated/>
      </Provider>,
    );
    const props = component.find('ProfilePage').props();
    expect(props.id).toBe(storeWithID.profile.id);
  });

  it('dispatches getProfile request', () => {
    const storeState = {
      profile: {
        id: 0,
      },
    };
    const store = mockStore(storeState);
    const sampleProfile = {
      id: 3423,
    };
    jest.spyOn(fakeApi, 'getProfile').mockImplementation(() => {
      return new Promise((resolve, reject) => (resolve({
        data: sampleProfile,
      })),
      );
    });
    const component = mount(
      <Provider store={store}>
        <ProfilePageDecorated/>
      </Provider>,
    );
    expect(fakeApi.getProfile).toBeCalledTimes(1);
  });
});
