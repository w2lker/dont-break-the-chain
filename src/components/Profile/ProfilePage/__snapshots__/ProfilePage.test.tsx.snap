// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePage component matches snapshot 1`] = `
"<Provider store={{...}}>
  <ProfilePage id={123} getProfile={[Function: mockConstructor]}>
    <WithStyles(undefined)>
      <Connect(ProfileHeader) classes={{...}}>
        <ProfileHeader classes={{...}} logout={[Function: logout]}>
          <div className=\\"Connect(ProfileHeader)-root-1\\">
            <h3 className=\\"Connect(ProfileHeader)-title-2\\">
              Profile
            </h3>
            <WithStyles(ForwardRef(IconButton)) className=\\"Connect(ProfileHeader)-button-3\\" onClick={[Function: logout]}>
              <ForwardRef(IconButton) classes={{...}} className=\\"Connect(ProfileHeader)-button-3\\" onClick={[Function: logout]}>
                <WithStyles(ForwardRef(ButtonBase)) className=\\"MuiIconButton-root Connect(ProfileHeader)-button-3\\" centerRipple={true} focusRipple={true} disabled={false} onClick={[Function: logout]}>
                  <ForwardRef(ButtonBase) classes={{...}} className=\\"MuiIconButton-root Connect(ProfileHeader)-button-3\\" centerRipple={true} focusRipple={true} disabled={false} onClick={[Function: logout]}>
                    <button className=\\"MuiButtonBase-root MuiIconButton-root Connect(ProfileHeader)-button-3\\" onBlur={[Function]} onClick={[Function: logout]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={0} type=\\"button\\" disabled={false}>
                      <span className=\\"MuiIconButton-label\\">
                        <MaterialDesignIcon name=\\"logout\\" className=\\"header-icon\\">
                          <span className=\\"material-icon mdi mdi-logout header-icon\\" style={[undefined]} />
                        </MaterialDesignIcon>
                      </span>
                      <NoSsr>
                        <WithStyles(undefined) center={true}>
                          <ForwardRef(TouchRipple) classes={{...}} center={true}>
                            <span className=\\"MuiTouchRipple-root\\">
                              <TransitionGroup component={{...}} exit={true} childFactory={[Function: childFactory]} />
                            </span>
                          </ForwardRef(TouchRipple)>
                        </WithStyles(undefined)>
                      </NoSsr>
                    </button>
                  </ForwardRef(ButtonBase)>
                </WithStyles(ForwardRef(ButtonBase))>
              </ForwardRef(IconButton)>
            </WithStyles(ForwardRef(IconButton))>
          </div>
        </ProfileHeader>
      </Connect(ProfileHeader)>
    </WithStyles(undefined)>
    <WithStyles(undefined)>
      <Connect(ProfileInfo) classes={{...}}>
        <ProfileInfo classes={{...}} avatar=\\"./img.png\\" name=\\"Test User\\" email=\\"tester@gmail.com\\" dispatch={[Function]}>
          <div className=\\"Connect(ProfileInfo)-wrapper-16\\">
            <div className=\\"Connect(ProfileInfo)-avatar-17\\" style={{...}} />
            <h2 className=\\"Connect(ProfileInfo)-name-18\\">
              Test User
            </h2>
            <p className=\\"Connect(ProfileInfo)-email-19\\">
              tester@gmail.com
            </p>
          </div>
        </ProfileInfo>
      </Connect(ProfileInfo)>
    </WithStyles(undefined)>
    <WithStyles(undefined)>
      <Connect(ProfileStats) classes={{...}}>
        <ProfileStats classes={{...}} total={245} days={12} dispatch={[Function]}>
          <div className=\\"Connect(ProfileStats)-wrapper-20\\">
            <div className=\\"Connect(ProfileStats)-statWrapper-21\\">
              <h4 className=\\"Connect(ProfileStats)-statValue-22\\">
                245
              </h4>
              <p className=\\"Connect(ProfileStats)-statDescription-24\\">
                points received
              </p>
            </div>
            <div className=\\"Connect(ProfileStats)-statWrapper-21\\">
              <h4 className=\\"Connect(ProfileStats)-statValue-22\\">
                12
                Â 
                <span className=\\"Connect(ProfileStats)-statSubtitle-23\\">
                  days
                </span>
              </h4>
              <p className=\\"Connect(ProfileStats)-statDescription-24\\">
                longest chain
              </p>
            </div>
          </div>
        </ProfileStats>
      </Connect(ProfileStats)>
    </WithStyles(undefined)>
  </ProfilePage>
</Provider>"
`;
